DataDog/dd-trace-rb
# [PROF-10241] Extract libdatadog crashtracker telemetry into separate extension

**What does this PR do?**

In #3384 we added support for the libdatadog crashtracker telemetry, and in #3816 we enabled it by default.

But so far the crashtracker telemetry support has been tied to the profiler:

1. It interfaces with libdatadog via the profiler native extension
2. It's only turned on when the profiler is turned on

This PR is the first step into allowing the crashtracker to be used completely independently of the profiler. Specifically, it addresses only 1) above, and does not yet address 2. (See additional notes for more details).

This PR introduces a new, separate native extension (`libdatadog_api`) that links to libdatadog separate from the profiler native extension, and moves the `crashtracker.c` bindings for the crashtracker to it.

This new extension borrows the same strategy as the profiler native extension for building and linking to libdatadog. This led to a bunch of code being extracted (and a bit being copy-pasted) from the profiler native extension.

The below diff looks "scarier" than it is, because most of the diff is things being moved around. After opening the PR, I plan to do a quick pass of comments to facilitate review and indicate more clearly when things just moved around.

Note that the crashtracker is working with this PR -- it's just still depending on profiling being enabled as well.

**Motivation:**

A big motivation for this work is to enable the crashtracker to work for customers not using profiling (or even in setups where the profiling doesn't work).

**Additional Notes:**

_There's some copy-pasta 🍝:_

I chose to copy-paste the `datadog_ruby_common.h`/`datadog_ruby_common.c` as it's kinda hard (and I did spend some time on it) to get `extconf.rb` to read `.c` and `.h` files from other directories, and to add them as correct dependencies to the generated `Makefile`. If you do figure out how to do it, please share!

_What's missing for later PRs:_

1. The crashtracker is still only being turned on when the profiler is turned on. We'll need to extract it to be a top-level component, outside `Datadog::Profiling`.

2. The native bits still place the crashtracker in `Datadog::Profiling::Crashtracker`, it should be moved as well.

3. The crashtracker extension is built in development only when running the profiler test suite. This does not affect an actual regular installation.

4. The crashtracker is controlled by the `DD_PROFILING_EXPERIMENTAL_CRASH_TRACKING_ENABLED` / `c.profiling.advanced.experimental_crash_tracking_enabled` setting. We need to deprecate this setting and rename it to something more appropriate.

5. Configuration of the endpoint is currently reusing some helper code in `profiling/http_transport.rb` that was not yet extracted. 

_Things to keep in mind about libdatadog/this integration_:

1. Libdatadog only has binary builds for linux. This means, like the profiler native extension, this is a "soft" integration that is built to gracefully work with this:
    1. If, during installation, the extension detects that it cannot compile, it will just skip that step and still allow a successful installation.
    2. You can force the extension to not be built with `DD_NO_EXTENSION=true`.
    3. For testing, you can set `DD_FAIL_INSTALL_IF_MISSING_EXTENSION=true` to force installation to fail immediately. This is useful when testing/debugging if the installation is working or not.
    6. We wrap extension loading with a begin/rescue, and the code gracefully handles situations where the extension is not present.

2. I've added a document showing how you can develop natively on macOS with libdatadog.

**How to test the change?**

This snippet will enable profiling and crash Ruby:

```
DD_TRACE_DEBUG=true DD_PROFILING_ENABLED=true bundle exec ddprofrb exec ruby -e 'Process.kill("SEGV", Process.pid)'
```

You should see logging indicating that the crashtracker has been turned on, and if you have an active Datadog agent running, you'll see a crash report being submitted.
